paper/panbench.tex

```diff
@@ -124,9 +122,176 @@ \section{Results}
+  assign :: expr -> stmt
+
+class While expr stmt where
+  while :: expr -> stmt -> stmt
+
+class AndThen stmt where
+  andThen :: stmt -> stmt -> stmt
+\end{lstlisting}
+
+This style of language encoding is typically known as \emph{finally tagless}\cite{Carette-Kiselyov-Shan:2009},
+and is well-known technique. However, our encoding is a slight refinement of the usual tagless style.
+In particular, we restrict ourselves to a single class per production rule, whereas other
+tagless encodings often use a class per syntactic category. We will explore the reasoning behind
+this more fine-grained approach in later sections.
+
+\subsection{Designing The Language}
+\label{sec:language-design}
+
+\todo[inline]{Reed: Too informal, just brain dumping.}
+
+Now that we've fleshed out how we are going to encode our grammatical framework into our host language,
+it's time to design our idealized abstract grammar. All of our host languages largely agree on a subset of
+the grammar of non-binding terms: the difficulties lie with binding forms and top level definitions.
+
+The difficulties with binding forms are two-fold. The first problem is that a binder in a production-grade
+proof assistant caries a lot more information than just a variable name and a type. After a careful study
+of our four intended target languages, we classified the various binding features along 3 separate axes:
+binding arity, binding annotations, and binding modifiers.
+
+Binding arities and annotations relatively self-explanatory, and classify the number of names bound, along with
+the type of annotation allowed. After studying our four target languages, we noticed that all binding
+arities fell into one of three classes: \(n\)-ary, unary, or nullary. We can similarly characterise annotations
+into three categories: required, optional, or forbidden.
+
+On the other hand, binding modifiers require a bit more explanation. A binding modifier capture features like
+implicit arguments, which do not change the number of names bound nor their annotations, but rather how
+those bound names get treated by the rest of the system. Features like irrelevant arguments in Agda
+or quantities in Idris 2 behave similarly: they do not change the shape of the binding. Currently,
+\panbench only supports visibility-related modifiers, but we have designed the system so that it is
+easy to extend with new modifiers.
+
+\todo[inline]{Reed: Make table of examples!}
+
+The question of top-level defintions is much thornier, and there seems to be less agreement on how
+they ought to be structured. The largest source of divergence is type signatures. Languages that
+prioritize dependent pattern matching typically opt to have standalone type signatures:
+this allow for top-level pattern matches, which in turn makes it much easier to infer motives\cite{McBride-Mckinna:2004}.
+Conversely, languages oriented around tactics typically opt for in-line type signatures, which
+leads to more compact definitions.
```

Comment:
Is that a post-facto reconstruction, or is the reason more Ocaml-style vs Haskell-style syntax? Which in turn may well be related to top-level pattern-matching definitions, but certainly not tactics!